RETO LATAM - Mateo Aristizabal

Este reto se trato de optimizar funciones en cuanto a memoria y tiempo de ejecucion

Fue proporcionado un archivo tipo json con informacion de tweets, se desarrolla localmente en un notebook y se usara las librerias propuestas en el README.md oara evaluar los aspectos de memoria y tiempo de ejecucion, es un total de 6 ejercicios. Ademas se subira a un repositorio Github con las buenas practicas del Gitflow y sera portable, ya que se puede desplegar en cualquier nube sin nungun problema. Todos los archivos mencionados se encuentran en la carpeta src, la version de python trabajada fue la 3.9 y un requirements.txt para saber las versiones de las librerias que se uso.

In [1]:
file_path = "/Users/josemateoaristizabaldiaz/downloads/farmers-protest-tweets-2021-2-4.json"

Se define la variable donde se encuentra el JSON en nuestro ambiente local.

In [2]:
from src import q1_memory, q1_time, q2_memory, q2_time, q3_memory, q3_time

Se importan los archivos que estan dentro de la carpeta src para llamar las funciones y evaluarlas solamente.

In [3]:
%load_ext memory_profiler

In [4]:
q1_time.q1_time(file_path)

[(datetime.date(2021, 2, 12), 'RanbirS00614606'),
 (datetime.date(2021, 2, 13), 'MaanDee08215437'),
 (datetime.date(2021, 2, 17), 'rebelpacifist'),
 (datetime.date(2021, 2, 16), 'jot__b'),
 (datetime.date(2021, 2, 14), 'jot__b'),
 (datetime.date(2021, 2, 18), 'RaaJVinderkaur'),
 (datetime.date(2021, 2, 15), 'neetuanjle_nitu'),
 (datetime.date(2021, 2, 20), 'Preetm91'),
 (datetime.date(2021, 2, 23), 'MangalJ23056160'),
 (datetime.date(2021, 2, 19), 'Surrypuria')]

Llamar la funcion, regresa lo esperado segun el ejemplo proporcionado.

In [15]:
q1_memory.q1_memory(file_path)

[(datetime.date(2021, 2, 12), 'RanbirS00614606'),
 (datetime.date(2021, 2, 13), 'MaanDee08215437'),
 (datetime.date(2021, 2, 17), 'rebelpacifist'),
 (datetime.date(2021, 2, 16), 'jot__b'),
 (datetime.date(2021, 2, 14), 'jot__b'),
 (datetime.date(2021, 2, 18), 'RaaJVinderkaur'),
 (datetime.date(2021, 2, 15), 'neetuanjle_nitu'),
 (datetime.date(2021, 2, 20), 'Preetm91'),
 (datetime.date(2021, 2, 23), 'MangalJ23056160'),
 (datetime.date(2021, 2, 19), 'Surrypuria')]

In [6]:
%timeit q1_time.q1_time(file_path)

2.67 s ± 45.4 ms per loop (mean ± std. dev. of 7 runs, 1 loop each)


In [19]:
%memit q1_time.q1_time(file_path)

peak memory: 1380.73 MiB, increment: 965.25 MiB


In [20]:
%timeit q1_memory.q1_memory(file_path)

3.06 s ± 6.26 ms per loop (mean ± std. dev. of 7 runs, 1 loop each)


In [21]:
%memit q1_memory.q1_memory(file_path)

peak memory: 516.44 MiB, increment: 0.06 MiB


In [38]:
q2_time.q2_time(file_path)

[('🙏', 5049),
 ('😂', 3072),
 ('🚜', 2972),
 ('🌾', 2182),
 ('🇮🇳', 2086),
 ('🤣', 1668),
 ('✊', 1651),
 ('❤️', 1382),
 ('🙏🏻', 1317),
 ('💚', 1040)]

In [39]:
q2_memory.q2_memory(file_path)

[('🙏', 5049),
 ('😂', 3072),
 ('🚜', 2972),
 ('🌾', 2182),
 ('🇮🇳', 2086),
 ('🤣', 1668),
 ('✊', 1651),
 ('❤️', 1382),
 ('🙏🏻', 1317),
 ('💚', 1040)]

In [17]:
%timeit q2_time.q2_time(file_path)

9.49 s ± 51.2 ms per loop (mean ± std. dev. of 7 runs, 1 loop each)


In [22]:
%memit q2_time.q2_time(file_path)

peak memory: 470.30 MiB, increment: 2.91 MiB


In [23]:
%timeit q2_memory.q2_memory(file_path)

9.61 s ± 30.8 ms per loop (mean ± std. dev. of 7 runs, 1 loop each)


In [24]:
%memit q2_memory.q2_memory(file_path)

peak memory: 470.28 MiB, increment: 0.02 MiB


In [25]:
q3_time.q3_time(file_path)

[('narendramodi', 2261),
 ('Kisanektamorcha', 1836),
 ('RakeshTikaitBKU', 1639),
 ('PMOIndia', 1422),
 ('RahulGandhi', 1125),
 ('GretaThunberg', 1046),
 ('RaviSinghKA', 1015),
 ('rihanna', 972),
 ('UNHumanRights', 962),
 ('meenaharris', 925)]

In [26]:
q3_memory.q3_memory(file_path)

[('narendramodi', 2261),
 ('Kisanektamorcha', 1836),
 ('RakeshTikaitBKU', 1639),
 ('PMOIndia', 1422),
 ('RahulGandhi', 1125),
 ('GretaThunberg', 1046),
 ('RaviSinghKA', 1015),
 ('rihanna', 972),
 ('UNHumanRights', 962),
 ('meenaharris', 925)]

In [27]:
%timeit q3_time.q3_time(file_path)

2.32 s ± 2.96 ms per loop (mean ± std. dev. of 7 runs, 1 loop each)


In [28]:
%memit q3_time.q3_time(file_path)

peak memory: 470.45 MiB, increment: 0.00 MiB


In [29]:
%timeit q3_memory.q3_memory(file_path)

2.39 s ± 21.2 ms per loop (mean ± std. dev. of 7 runs, 1 loop each)


In [30]:
%memit q3_memory.q3_memory(file_path)

peak memory: 471.12 MiB, increment: 0.52 MiB


Diferencias entre ejecuciones:

| Function | Time | Peak Memory | Memory Increment |
|----------|------|-------------|------------------|
| q1_memory | 2.67 s ± 45.4 ms per loop (mean ± std. dev. of 7 runs, 1 loop each) | 1380.73 MiB | 965.25 MiB |
| q1_time | 3.06 s ± 6.26 ms per loop (mean ± std. dev. of 7 runs, 1 loop each) | 2520.23 MiB | 2051.48 MiB |
| q2_memory | 9.61 s ± 30.8 ms per loop (mean ± std. dev. of 7 runs, 1 loop each) | 470.28 MiB | 0.02 MiB |
| q2_time | 9.49 s ± 51.2 ms per loop (mean ± std. dev. of 7 runs, 1 loop each) | 470.30 MiB | 2.91 MiB |
| q3_memory | 2.39 s ± 21.2 ms per loop (mean ± std. dev. of 7 runs, 1 loop each) | 471.12 MiB | 0.52 MiB |
| q3_time | 2.32 s ± 2.96 ms per loop (mean ± std. dev. of 7 runs, 1 loop each) | 470.45 MiB | 0.00 MiB |

Analisis:
